python/cuml/random_projection/rproj.pyx

```diff
@@ -0,0 +1,392 @@
+    and is thus suitable for methods having this requirement.
+
+    Sparse random matrix is an alternative to dense random projection matrix (e.g. Gaussian)
+    that guarantees similar embedding quality while being much more memory efficient
+    and allowing faster computation of the projected data (with sparse enough matrices).
+    If we note 's = 1 / density' the components of the random matrix are
+    drawn from:
+      - -sqrt(s) / sqrt(n_components)   with probability 1 / 2s
+      -  0                              with probability 1 - 1 / s
+      - +sqrt(s) / sqrt(n_components)   with probability 1 / 2s
+
+    Example
+    ---------
+
+    .. code-block:: python
+        from cuml.random_projection import SparseRandomProjection
+        from sklearn.datasets.samples_generator import make_blobs
+        from sklearn.svm import SVC
+
+        # dataset generation
+        data, target = make_blobs(n_samples=800, centers=400, n_features=3000, random_state=0)
+
+        # model fitting
+        model = SparseRandomProjection(n_components=5, random_state=42).fit(data)
+
+        # dataset transformation
+        transformed_data = model.transform(data)
+
+        # classifier training
+        classifier = SVC(gamma=0.001).fit(transformed_data, target)
+
+        # classifier scoring
+        score = classifier.score(transformed_data, target)
+
+        # measure information preservation
+        print("Score: {}".format(score))
+
+    Output:
+
+    .. code-block:: python
+        Score: 1.0
+
+    Parameters
+    ----------
+
+    n_components : int (default = 'auto')
+        Dimensionality of the target projection space. If set to 'auto',
+        the parameter is deducted thanks to Johnson–Lindenstrauss lemma.
+        The automatic deduction make use of the number of samples and
+        the eps parameter.
+
+        The Johnson–Lindenstrauss lemma can produce very conservative
+        n_components parameter as it makes no assumption on dataset structure.
+
+    density : float in range ]0, 1] (default = 'auto')
+        Ratio of non-zero component in the random projection matrix.
+
+        If density = 'auto', the value is set to the minimum density
+        as recommended by Ping Li et al.: 1 / sqrt(n_features).        
+
+    eps : float (default = 0.1)
+        Error tolerance during projection. Used by Johnson–Lindenstrauss
+        automatic deduction when n_components is set to 'auto'.
+
+    dense_output : boolean (default = True)
+        If set to True transformed matrix will be dense otherwise sparse.
+
+    random_state : int (default = 42)
+        Seed used to initilize random generator
+
+    Attributes
+    ----------
+        gaussian_method : boolean
+            To be passed to base class in order to determine
+            random matrix generation method
+
+    Notes
+    ------
+        Inspired from sklearn's implementation : https://scikit-learn.org/stable/modules/random_projection.html
+
+    """
+
```

Comment:
We need pytests for end-to-end validation. I would test the dataframe and numpy inputs for sparse and dense options, compare the johnson lindenstrauss computation against sklearn, and validate that the distances are within the established epsilon.

Reply:
Yes, I forgot the pytests.